### DISCUSSION TITLE
[Enhancement] Update to Vue3.3 and use Generic Component to provide better ts inferences

### DISCUSSION THREAD

**Discussion by evanryuu** [element-plus/element-plus#14103] [RFC]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `All`

### English：
Currently, element-plus is using version 3.2 of Vue, which does not support generic components. Should it be upgraded to version 3.3 and rewrite some components?

For example, in the `select` component, there are Props like `multiple` and `multipleLimit`. Obviously, if `multiple` is not set, the `multiple-limit` parameter should not be passed to the component. Does this require the use of generic components?

After defining props like this:
![image](https://github.com/element-plus/element-plus/assets/55378595/ff8c845b-5e3b-44e5-91b1-19028ce58fe5)

The usage would be like this:
![image](https://github.com/element-plus/element-plus/assets/55378595/2aade714-6c62-48b1-a498-fcf2c2322900)

Similarly, the `calendar-change` event in the `date-picker` component only exists when the `type` is `daterange` or `datetimerange`, so it also requires the use of generic components to provide better emits hints.

### 中文：
目前element-plus使用的vue还是3.2的版本，不支持泛型组件，是否应该升到3.3并重写部分组件？

举个例子，`select` 组件中有个 multiple 和 multipleLimit 的Props，那很明显如果没有设置 multiple 的话，不应该给组件传入 multiple-limit这个参数。那这好像必须得要使用泛型组件才能做到？

像这样定义props之后
![image](https://github.com/element-plus/element-plus/assets/55378595/ff8c845b-5e3b-44e5-91b1-19028ce58fe5)

使用起来就是这样的效果
![image](https://github.com/element-plus/element-plus/assets/55378595/2aade714-6c62-48b1-a498-fcf2c2322900)

同样的还有 `date-picker` 的 `calendar-change` 事件，它只有当 `type` 为 `daterange` 或者是 `datetimerange` 的时候才存在，这时候也需要用到泛型组件给到更好的emits提示。

> Example Repo: 
> 1. [select-generic.vue](https://github.com/evanryuu/element-plus/blob/feat/generic-comp/packages/components/select/src/select-generic.vue#L326-L327)
> 2. [test-select.vue](https://github.com/evanryuu/element-plus/blob/feat/generic-comp/packages/components/select/__tests__/test-select.vue)

### Additional comments

_(empty)_

**Comment by purepear** [comment]
I very much agree that updating to Vue 3.3 and implementing Generic Components will be a huge win! I think the main advantage will be the scoped slots. For example: ElTableColumn slots could be properly typed which will help us avoid errors

```vue
<script setup lang="ts">
interface Item {
  title: string
  url: string
}

const items: Item[] = [
  { title: 'Item 1', url: 'http://example.com/item-1' },
  { title: 'Item 2', url: 'http://example.com/item-2' },
]
</script>

<template>
  <ElTable :data="items">
    <ElTableColumn prop="title" label="Title" />
    <ElTableColumn prop="link" label="Link">
      <template #default="{ row }">
        <a :href="row.link">Link</a>  <----- We could catch this issue with type safety (should be row.url instead of row.link)
      </template>
    </ElTableColumn>
  </ElTable>
</template>
```
